### PR TITLE
Add transient notices feature

### LINKED ISSUE #6689: Remote inbox system: automatically install extensions from wp.org when clicking in the inbox CTA
**Description:** in order to remove friction, and increase the remote inbox effectiveness, the remote system should be able to autoinstall extensions from wp.org

**Acceptance criteria:**
- [ ] Automatically install extensions when the user clicks in the main call to action of a remote inbox message

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #6689 

Adds the ability to add/remove transient notices from the server.

I'd also love to hear if there's enough reusability here to also eventually use pieces of this with CES in the future or if there are other features that need to be added.  cc @becdetat @ilyasfoo 

### Screenshots

Screen Shot 2021-04-15 at 1 46 41 PM

### Detailed test instructions:

1. Attempt to add notices:
```php
\Automattic\WooCommerce\Admin\Features\TransientNotices::add(
    array(
        'id'      => 'my-notice',
        'status'  => 'success',
        'content' => 'Some information to display to the user.',
        'options' => array(
            'actions' => array(
                array(
                    'label' => 'Click me',
                    'url'   => 'http://wordpress.com',
                )
            )
        )
    )
);
```
2. Use the plugin installer in https://github.com/woocommerce/woocommerce-admin/pull/6805 and make sure notices are shown.
3. Optionally, throw some errors in `src/API/Plugins` in `install_plugins` or `activate_plugins` to make sure error messages are shown.

### CODE REVIEW

**Review by moon0326 — COMMENTED** [review]
Nice!! This is going to be very useful. I think the backend code can be re-used in other areas including CES if we add `category` or `type` so that we can distinguish notices in the frontend and handle them separately if necessary.

This is just an idea, but it would be nice to have notices store. 

An example pseudo-code:

**CES code**

```
\Automattic\WooCommerce\Admin\Features\TransientNotices::add(
    array(
        'id'      => 'my-notice',
        'status'  => 'success',
        'type'    => 'ces',
        'content' => 'Some information to display to the user.',
        'options' => array(
            'actions' => array(
                array(
                    'label' => 'Click me',
                    'url'   => 'http://wordpress.com',
                )
            )
        )
    )
);
```

```
// fetches CES notices and handle them in CES code
const notices = select('transient-notices')->byType('ces'); 
```

**Global notices**

```
\Automattic\WooCommerce\Admin\Features\TransientNotices::add(
    array(
        'id'      => 'my-notice',
        'status'  => 'success',
        'type'    => 'global',
        'content' => 'Some information to display to the user.',
        'options' => array(
            'actions' => array(
                array(
                    'label' => 'Click me',
                    'url'   => 'http://wordpress.com',
                )
            )
        )
    )
);
```


```
// this is for global -- global notices will be handled automatically
const globalNotices = select('transient-notices')->byType('global');
globalNotices.forEach( ( notice ) => {
    dispatch( 'core/notices' ).createNotice(
        notice.status,
        notice.content,
        notice.options || {}
    );

});
```

File: src/Features/TransientNotices.php
Comment (ilyasfoo) [inline src/Features/TransientNotices.php:87]: I think this has a pitfall, if say the user did not end up viewing the notice (like disconnected halfway through or navigated out just before the page finished loading), the notices will get deleted.

In CES we clear the queue via a trigger in `updateOptions` in @wordpress/data AFAIK, I think we can implement a similar thing here

File: src/Features/TransientNotices.php
Comment (adrianduffell) [inline src/Features/TransientNotices.php:41]: How does this work with multiple admin users?

For a while I've suspected there's a bug with the CES queue where users might see notices generated by another user if they are both active in the admin a the same time.

I wonder if this should be stored in user metadata instead?

**Comment by ilyasfoo** [comment]
This feature is a really great addition, thanks @joshuatf 👍 !

I think one of the challenges to adopt this in CES (at least in our current use of [`createNotice`](https://github.com/woocommerce/woocommerce-admin/blob/f3639c10a564c34ad15ac008965a53ff5d99bd0a/packages/customer-effort-score/src/index.js#L49-L62)) would be the inability to attach a callback which we rely upon. Though I don't know if there are any easy ways around this 🤷 .

File: src/Features/TransientNotices.php
Comment (joshuatf) [reply src/Features/TransientNotices.php:87]: I saw this and thought that may be the reason for updating this client-side.  I can see an error occurring from another plugin which would then prevent the notice from ever being shown.

I think this is worthwhile to add to this PR.

File: src/Features/TransientNotices.php
Comment (joshuatf) [reply src/Features/TransientNotices.php:41]: That's a good question.  I suspect this is a little edge case in terms of timing, but certainly a bug that could happen and probably a bug that does exist among many similar actions that are dependent on the next page load.  

I'm thinking if we do this, we could handle in a follow-up.  Do you think making notifications user specific are worthwhile?  Same question in regards to page specificity (only display on certain pages)?

**Comment by joshuatf** [comment]
Thanks for all the fantastic feedback, @moon0326, @ilyasfoo, and @adrianduffell!  This is super helpful.

>I think the backend code can be re-used in other areas including CES if we add category or type so that we can distinguish notices in the frontend and handle them separately if necessary.

>I think one of the challenges to adopt this in CES (at least in our current use of createNotice) would be the inability to attach a callback which we rely upon. Though I don't know if there are any easy ways around this 🤷 .

I think these two comments are around the same concern if I'm reading this correctly, is that right?  I like @moon0326's workaround of using a group, though I do think we may be able to solve this with `SlotFill` which would allow specifying callbacks by ID and then transient notices can continue handling the queuing of notices.  What do y'all think?

**Comment by joshuatf** [comment]
I updated this so that the notices are dequeued client-side instead of server-side avoiding the potential to dismiss early.  I also consolidated the code into the existing `TransientNotices` component as I think having multiple places for this to live was getting a bit confusing.

I started with a `SlotFill` approach, but found that this doesn't prove very DRY for the amount of elements in the snackbar.  Added a quick filter in https://github.com/woocommerce/woocommerce-admin/pull/6809/commits/9d3d56ddc6b11a4cf181fee0a594c77889a1627d.  This let's us easily filter the queued notices and add any props.  Does this work okay for the use case in CES, @ilyasfoo @moon0326?

Here's a quick example of its use:

```js
addFilter(
	'woocommerce_admin_queued_notice_filter',
	'my-namespace',
	( notice ) => {
		if ( notice.id === 'my-notice' ) {
			return {
				...notice,
				options: {
					...notice.options,
					actions: [
						{
							label: 'Custom action',
							onClick: () => alert( 'Filterrific!' ),
						},
					],
				},
			};
		}
	}
);
```

**Comment by moon0326** [comment]
> I updated this so that the notices are dequeued client-side instead of server-side avoiding the potential to dismiss early. I also consolidated the code into the existing `TransientNotices` component as I think having multiple places for this to live was getting a bit confusing.
> 
> I started with a `SlotFill` approach, but found that this doesn't prove very DRY for the amount of elements in the snackbar. Added a quick filter in [9d3d56d](https://github.com/woocommerce/woocommerce-admin/commit/9d3d56ddc6b11a4cf181fee0a594c77889a1627d). This let's us easily filter the queued notices and add any props. Does this work okay for the use case in CES, @ilyasfoo @moon0326?
> 
> Here's a quick example of its use:
> 
> ```js
> addFilter(
> 	'woocommerce_admin_queued_notice_filter',
> 	'my-namespace',
> 	( notice ) => {
> 		if ( notice.id === 'my-notice' ) {
> 			return {
> 				...notice,
> 				options: {
> 					...notice.options,
> 					actions: [
> 						{
> 							label: 'Custom action',
> 							onClick: () => alert( 'Filterrific!' ),
> 						},
> 					],
> 				},
> 			};
> 		}
> 		return notice;
> 	}
> );
> ```

Thank you @joshuatf 👍 How soon do we need to merge this PR? If we can wait a bit, I can try to use the filter in CES codebase and see how it would work next week. I also want some feedback from @becdetat when she comes back next week.

File: src/Features/TransientNotices.php
Comment (adrianduffell) [reply src/Features/TransientNotices.php:41]: Yeah a follow-up sounds good 👍 

> Do you think making notifications user specific are worthwhile?

I think it would be nice to have. For CES at least, a user needs to dismiss the notice or interact with it for it to leave the queue. As such, if they do nothing then the notice stays in the queue and then I suspect other users could see it on their next page view depending on the timing. 

It's still just a theory I have so I'm happy to investigate it some more. I would just like to avoid creating any confusion where a user get a notice that is out of context for them.

> Same question in regards to page specificity (only display on certain pages)?

This might be a CES-specific thing that wouldn't be needed normally. It was put in place to make sure notices only get triggered from certain pages. For example, the product notice only shows after updating the product from the product edit page. We wanted to avoid showing a notice when the product was updated elsewhere, like WP CLI, importer, bulk-editing etc.

**Comment by joshuatf** [comment]
@moon0326 I think we can wait a bit.  @pmcpinto do we have any notes that need the ability to install plugins immediately?

**Comment by moon0326** [comment]
@joshuatf 

I've tried to use this PR in CES and these are what I've found so far.

tl;dr -- Changes in this PR can definitely be used in CES, but I'm not sure if it's worth refactoring CES at this point. That said, I think we can use changes in this PR going forward.

1. CES uses `core/notices2`. I had to update [this line](https://github.com/woocommerce/woocommerce-admin/blob/bdd8169328bf1c1d10466eabf9d024b6f8eb39ad/client/layout/transient-notices/index.js#L25) to use `core/notices2`
2. This [line](https://github.com/woocommerce/woocommerce-admin/blob/bdd8169328bf1c1d10466eabf9d024b6f8eb39ad/src/Features/TransientNotices.php#L76) errored for me. I wasn't sure what to run to generate the script file. I commented it out for testing in dev.
3. It would be nice to have `getNotice($id)` getter in `TransientNotices` class.

**Comment by pmcpinto** [comment]
> @pmcpinto do we have any notes that need the ability to install plugins immediately?

Not that I'm aware of, but we're going to have it during Cycle 12. We aim to use the remote inbox to target existing Woo users to install WCPay

File: src/Features/TransientNotices.php
Comment (becdetat) [reply src/Features/TransientNotices.php:41]: I agree that this would cause notices to bleed through to other users. IMO making it per-user is worthwhile. Customer Effort Score also had the requirement that the notice only be displayed once per _store_ rather than once per _user_, which is most likely why I was ok with saving it as an option rather than per-user.

**Comment by joshuatf** [comment]
Thanks for reviewing again @moon0326 

>CES uses core/notices2. I had to update this line to use core/notices2

Cool, I don't see any issue with updating this to `core/notices2`.

>This line errored for me. I wasn't sure what to run to generate the script file. I commented it out for testing in dev.

Not quite sure what might cause that.  I think if you run `npm start` and `composer install` those dependencies should be built.  cc @psealock in case I'm missing any steps.

>It would be nice to have getNotice($id) getter in TransientNotices class.

Good idea.  Added a `get( $notice_id )` and `get_queue()` method.

File: src/Features/TransientNotices.php
Comment (joshuatf) [reply src/Features/TransientNotices.php:41]: I went ahead and added a `user_id` option to notices.  Any notices with a `user_id` will be shown to that user.  Notices without any `user_id` can be viewed by any user.

Example with the `PluginsInstaller` updated to show this in action.

**Comment by psealock** [comment]
>This line errored for me. I wasn't sure what to run to generate the script file. I commented it out for testing in dev.

`npm start` and its build process should create all that is needed? If its still broken, something else is off. What screen were you looking at?

**Comment by jeffstieler** [comment]
I'm getting the following errors trying to test:

```
Warning: require(/srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-admin/dist/wp-admin-scripts/transient-notices.min.asset.php): failed to open stream: No such file or directory in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-admin/src/Features/TransientNotices.php on line 76

Fatal error: require(): Failed opening required '/srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-admin/dist/wp-admin-scripts/transient-notices.min.asset.php' (include_path='.:/usr/share/php') in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-admin/src/Features/TransientNotices.php on line 76
```

**Comment by joshuatf** [comment]
@jeffstieler Did you re-run `npm start` after checking out this branch?  Should create the necessary dependency files.

**Review by moon0326 — COMMENTED** [review]
> > This line errored for me. I wasn't sure what to run to generate the script file. I commented it out for testing in dev.
> 
> `npm start` and its build process should create all that is needed? If its still broken, something else is off. What screen were you looking at @moon0326?

On any pages as soon as you add a notice with the following code. I've doubled checked the `dist/wp-admin-scripts` after running `composer install` and `npm start`. There was no `transient-notices.min.asset.ph` file.

```
\Automattic\WooCommerce\Admin\Features\TransientNotices::add(
    array(
        'id'      => 'my-notice',
        'status'  => 'success',
        'content' => 'Some information to display to the user.',
        'options' => array(
            'actions' => array(
                array(
                    'label' => 'Click me',
                    'url'   => 'http://wordpress.com',
                )
            )
        )
    )
);
```

File: src/Features/TransientNotices.php
Comment (psealock) [inline src/Features/TransientNotices.php:124]: >On any pages as soon as you add a notice with the following code. I've doubled checked the dist/wp-admin-scripts after running composer install and npm start. There was no transient-notices.min.asset.ph file.

Here is the offending code. It looks like c/p from dependencies gathered for `wp-admin-scripts` when Webpack creates the script. But this is an array written directly into the global scope and has no dependencies.

```js
var wcQueuedNotices = [ ... ]
```
You can remove the above two lines to fix the error. cc @moon0326

File: src/Features/TransientNotices.php
Comment (joshuatf) [reply src/Features/TransientNotices.php]: I pushed a commit showing this in action, but I'd prefer to roll it back.  I'm not as big a fan of doing this when methods contain more than a few arguments because:

1. It becomes difficult to use and remember the order of arguments.
```
TransientNotices::add(
    array(
        'user_id' => get_current_user_id(),
        'id'      => 'plugin-installer-' . str_replace( ',', '-', $plugins ),
        'status'  => 'success',
        'content' => $message,
    )
);
TransientNotices::add(
    'plugin-installer-' . str_replace( ',', '-', $plugins ),
    $message,
    array(),
    'success',
    get_current_user_id()
);
```
Using the latter I managed to make mistakes in both the readme and plugin installer since I mixed up the order of arguments.

2. Introducing new variables requires providing defaults for unused arguments and the consumer knowing more about the method.  This also reduces backwards compatibility.

If we're only concerned about enforcing required fields, we can add checks with the array style argument to enforce fields.

File: src/Features/TransientNotices.php
Comment (joshuatf) [reply src/Features/TransientNotices.php]: Ah 🤦  I forgot I removed the wp admin script in favor of using the main WCA library.  Pushed a fix for this. cc @moon0326 @jeffstieler

File: client/layout/transient-notices/index.js
Comment (joshuatf) [reply client/layout/transient-notices/index.js:52]: I think this may have been in an older version of this branch.  This needed to be updated to:

```
const notice = applyFilters( QUEUED_NOTICE_FILTER, queuedNotice );
```

File: src/Features/TransientNotices.php
Comment (jeffstieler) [reply src/Features/TransientNotices.php]: +1 to rolling back this change - the associative array approach is more inline with [WP standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#self-explanatory-flag-values-for-function-arguments).

Comment (moon0326) [reply src/Features/TransientNotices.php]: @joshuatf sounds good 👍

File: src/Features/TransientNotices.php
Comment (moon0326) [reply src/Features/TransientNotices.php]: @joshuatf 👍 The change worked for me.

**Review by jeffstieler — DISMISSED** [review]
Marking as "needs feedback" - rolling back the individual arguments.

**Comment by joshuatf** [comment]
Thanks, @jeffstieler - dropped that commit from this PR.

File: client/layout/transient-notices/index.js
Comment (samueljseay) [reply client/layout/transient-notices/index.js:52]: @joshuatf sorry I'm coming in a little naive here, so I'll start with a question,

is it desired to support adding transient notices via the client side as well as the backend? or is this purely a "per page load" feature.

I'm just asking that because the way this useEffect works, you won't be able to get any new transient notices if they were somehow added via the client, it will show them on first mount, but any further changes won't be reflected. Just checking really what the desired behaviour is.

File: src/PluginsInstaller.php
Comment (becdetat) [inline src/PluginsInstaller.php:68]: An exception occurs on line 73 if the `plugins` parameter is empty (triggering `src/API/Plugins.php:224`), causing a error page, I don't know if that should be fixed in this PR though.

File: client/layout/transient-notices/index.js
Comment (becdetat) [reply client/layout/transient-notices/index.js:52]: Ping @joshuatf. I don't _think_ this would be an issue as one could just call `useDispatch( 'core/notices2' ).createNotice()` directly to create a notice, unless I'm completely misunderstanding things.

**Review by becdetat — APPROVED** [review]
This is working well for me and the code looks respectable. I just had a couple of questions but I'll approve this :-)

File: client/layout/transient-notices/index.js
Comment (samueljseay) [reply client/layout/transient-notices/index.js:52]: Yes I think thats correct. The feature describes itself as "Transient notices provide a way to display snackbar notices from the server." so this seems entirely a feature for displaying a notice from a single server page load only. Of course creating notices via another mechanism would still work. 

I guess its worth mentioning that this wouldn't work in a case where you wanted to raise a notice during an API call (ie we fetch some data and there is no page load) and expect it to be displayed on screen. I think thats probably an edge case and I can't imagine that needing to be done, but just calling it out anyway.

One thing thats kinda weird here though is how `queuedNotices` is a piece of state that can change. Whats the point of storing it as state and updating it in dequeue? It makes sense to update the option but maintaining state from a single page load that can't change with this `useEffect` being basically the equivalent of onMount means that you'll never care about whats in queuedNotices after the initial page load?

File: src/Features/TransientNotices.php
Comment (adrianduffell) [reply src/Features/TransientNotices.php]: Thanks and nice job on this @joshuatf!

**Comment by joshuatf** [comment]
Thanks for the feedback, @becdetat and @samueljseay!  I've made some changes to how the notices are fetched and wouldn't mind another look at this one before merging if you have time.

File: src/Features/TransientNotices.php
Comment (becdetat) [inline src/Features/TransientNotices.php]: Am I correct in thinking that this no longer enqueues notices for the current user? Would this introduce timing issues?

Comment (becdetat) [reply src/Features/TransientNotices.php]: Ah no, I can see in the next commit that the user check is happening on the client now. Carry on :-)

**Comment by becdetat** [comment]
Changes look good, and it all still works well. LGTM!

**Pull request merged**
